ouroboros-consensus/ouroboros-consensus-test-infra/src/Test/Util/HardFork/Future.hs

```diff
@@ -0,0 +1,151 @@
+  futureFirstSlotLength,
+  futureSlotLengths,
+  futureSlotToEpoch,
+  futureTimeToSlot,
+  singleEraFuture,
+  ) where
+
+import           Data.Time (NominalDiffTime)
+import           Data.Word (Word64)
+import           GHC.Generics (Generic)
+import           Quiet (Quiet (..))
+
+import           Cardano.Slotting.Slot (EpochNo (..), EpochSize (..),
+                     SlotNo (..))
+
+import           Ouroboros.Consensus.BlockchainTime (SlotLength (..))
+import           Ouroboros.Consensus.Util (nTimes)
+
+import           Test.Util.Slots (NumSlots (..))
+import           Test.Util.Stream (Stream (..))
+
+{-------------------------------------------------------------------------------
+  Careful counts
+-------------------------------------------------------------------------------}
+
+-- | Number of epochs
+newtype EraSize = EraSize {unEraSize :: Word64}
+  deriving (Eq, Generic)
+  deriving (Show) via (Quiet EraSize)
+
+{-------------------------------------------------------------------------------
+  A test's whole timeline
+-------------------------------------------------------------------------------}
+
+-- | Every era in the test
+--
+-- INVARIANT: every number is @> 0@
+data Future =
+      FinalEra SlotLength EpochSize
+    | NextEra  SlotLength EpochSize EraSize Future
+  deriving (Eq, Show)
+
+-- | 'Future' with only one era
+singleEraFuture :: SlotLength -> EpochSize -> Future
+singleEraFuture = FinalEra
+
+-- | 'SlotLength' of the first era
+futureFirstSlotLength :: Future -> SlotLength
+futureFirstSlotLength future = case future of
+    NextEra  slotLength _epochSize _eraSize _future -> slotLength
+    FinalEra slotLength _epochSize                  -> slotLength
+
+-- | 'EpochSize' of the first era
+futureFirstEpochSize :: Future -> EpochSize
+futureFirstEpochSize future = case future of
+    NextEra  _slotLength epochSize _eraSize _future -> epochSize
+    FinalEra _slotLength epochSize                  -> epochSize
+
+-- | Future as of a slot within an epoch
+data SlotFuture =
+      InFinalEra SlotLength
+    | InNextEra  SlotLength NumSlots Future
+      -- ^ INVARIANT: @'NumSlots' > 0@
+
+-- | Length of each slot in the whole 'Future'
+futureSlotLengths :: Future -> Stream SlotLength
+futureSlotLengths = \case
+    FinalEra slotLength _epochSize ->
+        let x = slotLength :< x in x
+    NextEra slotLength epochSize eraSize future ->
+        nTimes (slotLength :<) eraSlots $
+        futureSlotLengths future
+      where
+        NumSlots eraSlots = calcEraSlots epochSize eraSize
+
+-- | @(slot, time left in slot, length of slot)@
+futureTimeToSlot :: Future
+                 -> NominalDiffTime
+                 -> (SlotNo, NominalDiffTime, SlotLength)
+futureTimeToSlot = \future d -> go2 0 d future
+  where
+    go acc d (InFinalEra slotLength) =
+        (SlotNo $ acc + n, timeInSlot, slotLength)
+      where
+        n = divide d slotLength
+        timeInSlot = d - multiply n slotLength
```

Comment:
Suppose `d = 3.8`, `slotLength = 1.0`, then `n` will be `3`, and hence `timeInSlot = 3.8 - 3 = 0.8`. This is the time _spent_ in the lost, right, not the time _left_ in the slot. Is the comment above this function incorrect?

Reply:
Great catch; the comment was wrong. But I updated the code (use call sites) to match the comment; it's tidier.